app/controllers/course_summaries_controller.rb

```diff
@@ -15,4 +15,116 @@ def populate
   def get_marking_scheme_details
     redirect_to url_for(controller: 'marking_schemes', action: 'populate')
   end
+
+  def download_csv_grades_report
+    csv_string = CSV.generate do |csv|
+      # Populates the entire CSV with headers, students and their marks
+      populate_students_and_marks(csv)
+    end
+    course_name = "#{COURSE_NAME}"
+    course_name_underscore = course_name.squish.downcase.tr(' ', '_')
+    send_data csv_string, disposition: 'attachment',
+                          filename: "#{course_name_underscore}_grades_report.csv"
+  end
+end
+
+# ---- HELPER FUNCTIONS FOR DOWNLOAD_CSV_GRADES_REPORT BELOW-----#
+
+# Populates the headers in the CSV
+def populate_header_titles(assignments)
+  header = ['Username']
+  assignments.each do |assignment|
+    header.push(assignment.short_identifier)
+  end
+  header.push('Total marks')
+end
+
+# Checks grouping, submission and it's mark for nil.
+def fill_row_with_marks(grouping, row)
```

Comment:
Couple of comments:

1. This should either mutate `row` or return a new row, but not both.
2. The name is misleading, given that only one mark is ever pushed.

I suggest making this helper simply return either `''` or the mark, and actually pushing the mark in line 65.